src/nvim/eval/funcs.c

```diff
@@ -5404,8 +5404,17 @@ static void f_jobwait(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   TV_LIST_ITER_CONST(args, arg, {
     Channel *chan = NULL;
     if (TV_LIST_ITEM_TV(arg)->v_type != VAR_NUMBER
-        || !(chan = find_job(TV_LIST_ITEM_TV(arg)->vval.v_number, false))) {
+        || !(chan = find_channel(TV_LIST_ITEM_TV(arg)->vval.v_number))
+        || chan->streamtype != kChannelStreamProc) {
       jobs[i] = NULL;  // Invalid job.
+    } else if (process_is_stopped(&chan->stream.proc)) {
+      // Process is already stopped, so ensure all callbacks on its event
+      // queue are drained
+      Process *proc = &chan->stream.proc;
+      LOOP_PROCESS_EVENTS_UNTIL(proc->loop, proc->events, -1, !proc->refcount);
```

Comment:
- any reason not to use `process_wait()`? that avoids reaching into `proc->refcount` (informally private), and also fits semantically, so a reader can see that process_wait() is called in multiple places in jobwait().

```suggestion
      process_wait(&chan->stream.proc, -1, NULL);
```

also via @bfredl:

> it should be documented that any object holding a proc->refcount must release that eventually by an event on the process own event queue. and i.e. not by some event on some other child queue (only)

Reply:
The idea was to cut to the chase and sidestep the other stuff that process_wait does, as well as being explicit that we are just processing the event queue and not necessarily “waiting” for a process that has already stopped.

But I think you’re right that it may be clearer. And I wasn’t aware of the semantics around `refcount`, so I think your suggestion makes sense.

(Btw I’m replying on mobile and just using the “Commit suggestion” box, so forgive the lack of better commit messages. I figure they can be squashed/rebased later).